core/src/ibc/channel_04/manager.rs

```diff
@@ -40,4 +44,61 @@ impl<'a> Manager<'a> {
             ctx,
         }
     }
+
+    pub fn chan_open_init(
+        &mut self,
+        order: ChannelOrder,
+        connection: Identifier,
+        channel_identifier: Identifier,
+        counterparty_channel_identifier: Identifier,
+        version: String,
+    ) -> Result<Identifier, String> {
+        let kv_store = self.ctx.get_kv_store_mut();
+
+        // It is ok to be in any state.
+        let _: ConnectionEnd = rlp::decode(
+            &kv_store.get(&connection_path(&connection)).ok_or_else(|| "Connection doesn't exist".to_owned())?,
+        )
+        .expect("Illformed connection end stored in the DB");
+
+        let channel = ChannelEnd {
+            state: ChannelState::INIT,
+            ordering: order,
+            counterparty_port_identifier: DEFAULT_PORT.to_string(),
+            counterparty_channel_identifier,
+            connection_hops: vec![connection],
+            version,
+        };
+
+        // NOTE: You don't have to recover the state even if the transaction fails.
+        // NOTE2: Is it ok to rewrite ChannelEnd when it exists? Currently it fails the transaction.
```

Comment:
Please write concerns in an issue, not in the comment. We can't discuss using comments.

Reply:
Please write a result of a discussion in a comment.